analytic_engine/src/memtable/layered/mod.rs

```diff
@@ -0,0 +1,459 @@
+    }
+
+    pub fn min_key(&self) -> Option<Bytes> {
+        let comparator = BytewiseComparator;
+
+        let mut mutable_min_key = self.mutable_segment.min_key();
+
+        let immutable_min_key = if self.immutable_segments.is_empty() {
+            None
+        } else {
+            let mut min_key = self.immutable_segments.first().unwrap().min_key();
+            let mut imm_iter = self.immutable_segments.iter();
+            let _ = imm_iter.next();
+            for imm in imm_iter {
+                match comparator.compare_key(&min_key, &imm.min_key) {
+                    std::cmp::Ordering::Greater => min_key = imm.min_key(),
+                    std::cmp::Ordering::Less | std::cmp::Ordering::Equal => (),
+                }
+            }
+
+            Some(min_key)
+        };
+
+        match (mutable_min_key, immutable_min_key) {
+            (None, None) => None,
+            (None, Some(key)) | (Some(key), None) => Some(key),
+            (Some(key1), Some(key2)) => Some(match comparator.compare_key(&key1, &key2) {
+                std::cmp::Ordering::Greater => key2,
+                std::cmp::Ordering::Less | std::cmp::Ordering::Equal => key1,
+            }),
+        }
+    }
+
+    pub fn max_key(&self) -> Option<Bytes> {
+        let comparator = BytewiseComparator;
+
+        let mut mutable_max_key = self.mutable_segment.max_key();
+
+        let immutable_max_key = if self.immutable_segments.is_empty() {
+            None
+        } else {
+            let mut max_key = self.immutable_segments.first().unwrap().max_key();
+            let mut imm_iter = self.immutable_segments.iter();
+            let _ = imm_iter.next();
+            for imm in imm_iter {
+                match comparator.compare_key(&max_key, &imm.max_key) {
+                    std::cmp::Ordering::Less => max_key = imm.max_key(),
+                    std::cmp::Ordering::Greater | std::cmp::Ordering::Equal => (),
+                }
+            }
+
+            Some(max_key)
+        };
+
+        match (mutable_max_key, immutable_max_key) {
+            (None, None) => None,
+            (None, Some(key)) | (Some(key), None) => Some(key),
+            (Some(key1), Some(key2)) => Some(match comparator.compare_key(&key1, &key2) {
+                std::cmp::Ordering::Less => key2,
+                std::cmp::Ordering::Greater | std::cmp::Ordering::Equal => key1,
+            }),
+        }
+    }
+
+    pub fn time_range(&self) -> Option<TimeRange> {
+        let mutable_time_range = self.mutable_segment.time_range();
+
+        let immutable_time_range = if self.immutable_segments.is_empty() {
+            None
+        } else {
+            let mut time_range = self.immutable_segments.first().unwrap().time_range();
+            let mut imm_iter = self.immutable_segments.iter();
+            let _ = imm_iter.next();
+            for imm in imm_iter {
+                time_range = time_range.intersected_range(imm.time_range()).unwrap();
```

Comment:
Will the intersection of all the time ranges make a smaller time range?

Reply:
It should merge to expand the time range, has been fixed.